subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/CollectDirectClassSuperTypesTransform.java

```diff
@@ -0,0 +1,124 @@
+import java.io.UncheckedIOException;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import static com.google.common.collect.ImmutableSortedSet.toImmutableSortedSet;
+
+/**
+ * TODO: This class has similar implementation in build-logic/packaging/src/main/kotlin/gradlebuild/instrumentation/transforms/CollectDirectClassSuperTypesTransform.kt.
+ *  We could reuse the same class at some point.
+ */
+@DisableCachingByDefault(because = "Not worth caching.")
+public abstract class CollectDirectClassSuperTypesTransform implements TransformAction<TransformParameters.None> {
+
+    private static final Predicate<String> ACCEPTED_TYPES = type -> type != null && !type.startsWith("java/lang");
+    public static final String SUPER_TYPES_MARKER_FILE_NAME = ".gradle-super-types.marker";
+    private static final String FILE_SUFFIX = ".super-types";
+
+    @Inject
+    public abstract ObjectFactory getObjects();
+
+    @PathSensitive(PathSensitivity.NAME_ONLY)
+    @InputArtifact
+    public abstract Provider<FileSystemLocation> getInput();
+
+    public void transform(TransformOutputs outputs) {
+        try {
+            InjectedInstrumentationServices services = getObjects().newInstance(InjectedInstrumentationServices.class);
+            ClasspathWalker walker = services.getClasspathWalker();
+            File inputFile = getInput().get().getAsFile();
+            Map<String, Set<String>> superTypes = new TreeMap<>();
+            try {
+                walker.visit(inputFile, entry -> {
+                    if (entry.getName().endsWith(".class")) {
+                        ClassReader reader = new ClassReader(entry.getContent());
+                        String className = reader.getClassName();
+                        Set<String> classSuperTypes = getSuperTypes(reader);
+                        if (!classSuperTypes.isEmpty()) {
+                            superTypes.put(className, classSuperTypes);
+                        }
+                    }
+                });
+            } catch (FileException zipException) {
+                // We support badly formatted jars on the build classpath
+                // see: https://github.com/gradle/gradle/issues/13816
+                return;
+            }
+
+            File output = outputs.file(inputFile.getName() + FILE_SUFFIX);
+            try (BufferedWriter writer = new BufferedWriter(new FileWriter(output))) {
+                for (Map.Entry<String, Set<String>> entry : superTypes.entrySet()) {
+                    writer.write(entry.getKey() + "=" + String.join(",", entry.getValue()) + "\n");
+                }
+            }
+
+            // Mark the folder so we know that this is a folder with super types files
+            // This is currently used just to not delete the folders for performance testing
+            outputs.file(SUPER_TYPES_MARKER_FILE_NAME).createNewFile();
```

Comment:
❔ Why do we need to not delete these folders for performance testing?

Reply:
We have performance [tests that clears artifact transform tests](https://github.com/gradle/gradle/blob/f8ee6cae5bd056483bafec72043744540f532e09/subprojects/performance/src/performanceTest/groovy/org/gradle/performance/regression/android/RealLifeAndroidBuildPerformanceTest.groovy#L91). 

And if we clear instrumentation artifact transforms, we get a performance regression compared to previous versions, since we run also instrumentation artifact transforms.